src/daft-scan/src/lib.rs

```diff
@@ -1,93 +1,318 @@
 };
 
-use common_error::{DaftError, DaftResult};
+use common_error::DaftResult;
 use daft_core::{datatypes::Field, schema::SchemaRef};
-use daft_dsl::Expr;
+use daft_dsl::{Expr, ExprRef};
 use daft_stats::{PartitionSpec, TableMetadata, TableStatistics};
+use file_format::FileFormatConfig;
 use serde::{Deserialize, Serialize};
 
 mod anonymous;
+pub mod file_format;
+#[cfg(feature = "python")]
+pub mod py_object_serde;
+
 #[cfg(feature = "python")]
 pub mod python;
+pub mod storage_config;
 #[cfg(feature = "python")]
 pub use python::register_modules;
+use storage_config::StorageConfig;
 
-#[derive(Serialize, Deserialize, Clone, Copy, Debug)]
-pub enum FileType {
-    Parquet,
-    Avro,
-    Orc,
-    Csv,
-}
-
-impl FromStr for FileType {
-    type Err = DaftError;
-
-    fn from_str(file_type: &str) -> DaftResult<Self> {
-        use FileType::*;
-        if file_type.trim().eq_ignore_ascii_case("parquet") {
-            Ok(Parquet)
-        } else if file_type.trim().eq_ignore_ascii_case("avro") {
-            Ok(Avro)
-        } else if file_type.trim().eq_ignore_ascii_case("orc") {
-            Ok(Orc)
-        } else if file_type.trim().eq_ignore_ascii_case("csv") {
-            Ok(Csv)
-        } else {
-            Err(DaftError::TypeError(format!(
-                "FileType {} not supported!",
-                file_type
-            )))
-        }
-    }
-}
-
-#[derive(Serialize, Deserialize)]
+#[derive(Debug, Clone, Serialize, Deserialize)]
 pub enum DataFileSource {
     AnonymousDataFile {
-        file_type: FileType,
         path: String,
         metadata: Option<TableMetadata>,
         partition_spec: Option<PartitionSpec>,
         statistics: Option<TableStatistics>,
     },
     CatalogDataFile {
-        file_type: FileType,
         path: String,
         metadata: TableMetadata,
         partition_spec: PartitionSpec,
         statistics: Option<TableStatistics>,
     },
 }
 
-#[derive(Serialize, Deserialize)]
+impl DataFileSource {
+    pub fn get_path(&self) -> &str {
+        match self {
+            Self::AnonymousDataFile { path, .. } | Self::CatalogDataFile { path, .. } => path,
+        }
+    }
+    pub fn get_metadata(&self) -> Option<&TableMetadata> {
+        match self {
+            Self::AnonymousDataFile { metadata, .. } => metadata.as_ref(),
+            Self::CatalogDataFile { metadata, .. } => Some(metadata),
+        }
+    }
+
+    pub fn get_statistics(&self) -> Option<&TableStatistics> {
+        match self {
+            Self::AnonymousDataFile { statistics, .. }
+            | Self::CatalogDataFile { statistics, .. } => statistics.as_ref(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct ScanTask {
     // Micropartition will take this in as an input
-    source: DataFileSource,
-    columns: Option<Vec<String>>,
-    limit: Option<usize>,
+    pub source: DataFileSource,
+    pub file_format_config: Arc<FileFormatConfig>,
+    pub schema: SchemaRef,
```

Comment:
We could maybe build a stronger abstraction around `ScanTask` so that it is much lighter-weight, and merely borrows data from its parent `ScanTaskBatch`.

Single `ScanTasks` will just be wrapped in a `ScanTaskBatch` of size 1.

```
struct ScanTaskBatch {...}

impl ScanTaskBatch {
    fn scan_tasks<'a>(&'a self) -> Iterator<ScanTask<'a>> {...}
}

/// ScanTask is just a "view" on some of the data in its parent ScanTaskBatch and
/// cannot really exist without the parent ScanTaskBatch
struct ScanTask<'batch> {
    path: &'batch str,
    format: &'batch FileFormatConfig,
    schema: &'batch SchemaRef,
    ...
}
```

This way, the `ScanTaskBatch` owns all the data, and the `ScanTask` is just a lightweight view on whatever data is presented by the `ScanTaskBatch`, which can then have full control around de-duping information where necessary.

Reply:
I played around with it a little -- turns out we actually don't use `ScanTask` for anything other than returning a `Vec<ScanTask>` from our ScanOperators, and constructing a `ScanTaskBatch`. From there onwards, it's actually all just `ScanTaskBatch` everywhere 😛 

I.e. it was actually very safe to completely remove the `ScanTask` abstraction, and just have `ScanTaskBatch`. Another way to think about it is that our scan tasks inherently are multi-file, and batched (we just have it hardcoded to batches of size 1 at the moment), so maybe we should rename `ScanTaskBatch -> ScanTask` and call it a day.

Here's a PR to show what that looks like, requiring surprisingly few changes: https://github.com/Eventual-Inc/Daft/pull/1565